Move CI and IDE metadata processing at execution time

jprinet opened the pull request:
The PR aims to capture CI and IDE metadata at execution time (in a `buildFinished` block), therefore preventing configuration cache invalidation if a property changes (ie. `CI build number`)

Teamcity data are now fetched from a teamcity generated build file as project properties set in an init file can't easily be read at execution time through a [Provider](https://docs.gradle.org/current/javadoc/org/gradle/api/provider/ProviderFactory.html#gradleProperty-java.lang.String-) due to https://github.com/gradle/gradle/issues/23572.

The `projectEvaluated` callback has been replaced by `Providers` used at execution time in order to fix configuration cache restore scenarii.

Note that Teamcity is not yet compatible with configuration cache due to [this](https://youtrack.jetbrains.com/issue/TW-71916)

fixes https://github.com/gradle/common-custom-user-data-gradle-plugin/issues/80

This PR was tested with 
- Gradle 6.0, 7.2, 7.6
- w/o configuration cache
- Jenkins 2.332.3
- Teamcity 2022.10.1 (build 116934)
- Android Studio 2021.3.1 Patch 1

Signed-off-by: Jerome Prinet [jprinet@gradle.com](mailto:jprinet@gradle.com)

runningcode commented on src/main/java/com/gradle/CustomBuildScanEnhancements.java:
```diff
@@ -55,215 +64,289 @@ void apply() {
-            }));
+        if (CiUtils.isCi(providers)) {
+            // Prepare project directory for use at execution time
+            Provider<Directory> projectDirectory = providers.provider(() -> gradle.getRootProject().getLayout().getProjectDirectory());
```
is this Gradle 4 or 5 compatible?

runningcode commented on src/main/java/com/gradle/CustomBuildScanEnhancements.java:
```diff
@@ -55,215 +64,289 @@ void apply() {
 
-            buildId.ifPresent(value ->
-                buildScan.value("CI build number", value));
+        private static Properties readPropertiesFile(String fileName, ProviderFactory providers, Directory projectDirectory) {
```
what does this method declaration do? it seems like just a redirect

jprinet replied:
Absolutely, this was already there before the PR
https://github.com/gradle/common-custom-user-data-gradle-plugin/blob/97e48144a0d4370d3e2bb409bcba02878adb4aa7/src/main/java/com/gradle/CustomBuildScanEnhancements.java#L527

jprinet replied:
Just removed the indirection, that's better now 👍

runningcode reviewed (commented):
I know it isn't ready for review but I added two comments. When the PR is ready, please add what versions of Gradle and TeamCity this was tested against :)

runningcode commented on src/main/java/com/gradle/CustomBuildScanEnhancements.java:
```diff
@@ -60,210 +69,280 @@ private void captureOs() {
-            Utils.envVariable("CI_JOB_STAGE", providers).ifPresent(value ->
-                customValueSearchLinker.addCustomValueAndSearchLink("CI stage", value));
-        }
+            if(CiUtils.isTeamCity(providers)) {
```
```suggestion
            if (CiUtils.isTeamCity(providers)) {
```
let's fix the formatting from the previous PR

runningcode reviewed (commented):
another round of comments. looking good. i just wanted to leave these comments but I haven't finished the review yet

runningcode commented on src/main/java/com/gradle/CustomBuildScanEnhancements.java:
```diff
@@ -60,210 +69,280 @@ private void captureOs() {
+                    }
+                    String teamCityAgentName = buildProperties.getProperty("agent.name");
+                    if (Utils.isNotEmpty(teamCityAgentName)) {
+                        buildScan.value("CI agent", teamCityAgentName);
```
before we were calling the `customValueSearchLinker` why did we change it here? Is it not needed?

jprinet replied:
Well spotted, this one just slipped through the cracks while refactoring, thanks!

runningcode reviewed (commented):
Can you also test this with Gradle 4 and 5?
Since TC is not yet CC compatible, did you manually test the TC configuration some other way to make sure the configuration cache was re-usable?
Ok, done for this round of reviews.

jprinet commented on src/main/java/com/gradle/CustomBuildScanEnhancements.java:
```diff
@@ -60,210 +69,280 @@ private void captureOs() {
+        if(Utils.isGradle61OrNewer()) {
+            return providers.systemProperty(name);
+        } else {
+            return providers.provider(() -> System.getProperty(name));
```
The condition is not about deprecation but availability of the method. 
[systemProperty](https://docs.gradle.org/current/javadoc/org/gradle/api/provider/ProviderFactory.html#systemProperty-java.lang.String-) was introduced in 6.1

runningcode replied:
Is there any reason we cannot use the following code for all Gradle versions?
```
providers.provider(() -> System.getProperty(name))
```

jprinet replied:
We surely could, I prefer to have the idiomatic way clearly highlighted with a condition on Gradle version to be able to get rid of legacy way once we decide not to support Gradle < 6.1.

runningcode replied:
I see. This seems like an interesting trade off of best practices vs code simplicity. Let's keep it but can we move this to the bottom of the file in a static method where the other methods are? (along with the `gradlePropertyProvider`)

runningcode replied:
sounds good.

jprinet commented on src/main/java/com/gradle/CustomBuildScanEnhancements.java:
```diff
@@ -60,210 +69,280 @@ private void captureOs() {
-                    buildScan.tag("Cmd Line");
-                }
+            // Prepare relevant properties for use at execution time
+            Map<String, Provider<String>> relevantProperties = new HashMap<>();
```
Credit goes to @bamboo there

jprinet commented on src/main/java/com/gradle/CustomBuildScanEnhancements.java:
```diff
@@ -60,210 +69,280 @@ private void captureOs() {
-                }
-            });
+    private Provider<String> systemPropertyProvider(String name, ProviderFactory providers) {
+        if(Utils.isGradle61OrNewer()) {
```
I reformatted the code and stop ignored whitespaces in my commit view

runningcode commented on src/main/java/com/gradle/CustomBuildScanEnhancements.java:
```diff
@@ -60,210 +69,289 @@ private void captureOs() {
+        if (Utils.isGradle62OrNewer()) {
+            return providers.gradleProperty(name);
+        } else {
+            return providers.provider(() -> (String) gradle.getRootProject().findProperty(name));
```
is there any reason we cannot the following for all Gradle versions?
`providers.provider(() -> (String) gradle.getRootProject().findProperty(name))`

jprinet replied:
same reason as `providers.systemProperty`

jprinet commented:
> Can you also test this with Gradle 4 and 5? Since TC is not yet CC compatible, did you manually test the TC configuration some other way to make sure the configuration cache was re-usable? Ok, done for this round of reviews.

I tested with Gradle 4.1 and 5.
It was a good call as [getOrElse](https://docs.gradle.org/current/javadoc/org/gradle/api/provider/Provider.html#getOrElse-T-) was introduced in 4.3, I added a condition to enable 4.1 compatibility.

Regarding Teamcity use case, the only data read is an environment variable at execution time. Jenkins case is processed the same way and was successfully tested with configuration cache enabled.

runningcode commented on src/main/java/com/gradle/CustomBuildScanEnhancements.java:
```diff
@@ -55,215 +71,294 @@ void apply() {
+
+                    String teamcityConfigFile = buildProperties.getProperty("teamcity.configuration.properties.file");
+                    if (isNotEmpty(teamcityConfigFile)) {
+                        Properties configProperties = readPropertiesFile(teamcityConfigFile, providers, projectDirectory.get());
```
In the previous version we would only read the file if the build id was also present. why did we change this? I'm not saying either way is wrong or correct, I just want to understand.

jprinet replied:
I found it easier to read and assumption is that build id is always present.
Let's revert to the original approach though

runningcode replied:
I didn't mean to suggest anything, I was just curious if there was a reason behind it. :)

runningcode commented on src/main/java/com/gradle/CustomBuildScanEnhancements.java:
```diff
@@ -55,215 +71,294 @@ void apply() {
-                }
-            });
+            // Prepare relevant properties for use at execution time
+            Map<String, Provider<String>> relevantProperties = new HashMap<>();
```
can we name this variable `ideProperties`
```suggestion
            Map<String, Provider<String>> ideProperties = new HashMap<>();
```

runningcode reviewed (commented):
Ok, I've added my review. :)
Looking good. I like the static import changes. I think we should be consistent about their uses though. Let's either always use them or never use them. In this case I think we should always use them.
Can we double check that on TeamCity (and spot check some other CI systems) that all the links in the build scan appear the same as in the previous released version of CCUD? Please also add a build scan link in the PR with the before and after.

jprinet commented:
> Ok, I've added my review. :) Looking good. I like the static import changes. I think we should be consistent about their uses though. Let's either always use them or never use them. In this case I think we should always use them. Can we double check that on TeamCity (and spot check some other CI systems) that all the links in the build scan appear the same as in the previous released version of CCUD? Please also add a build scan link in the PR with the before and after.

I aligned the static imports as suggested 👍

Thanks as well for asking to check the navigation links, I realized that those were not displayed due to an overlooked [execution in buildFinished block](https://github.com/gradle/common-custom-user-data-gradle-plugin/blob/a5589b79b04ee9fff6dc2a519409c7b60aa64a9f/src/main/java/com/gradle/CustomBuildScanEnhancements.java#L372).
I adjusted the code to now process the links in sequence ([related commit](https://github.com/gradle/common-custom-user-data-gradle-plugin/pull/123/commits/49cfb0d730caa4253c36529df9b24b05d05d93c3)). I also decided to have `captureCiMetadata` and `captureGitMetadata` to each have its own instance of the `CustomValueSearchLinker` to simplify process ordering.

runningcode commented on src/main/java/com/gradle/CustomBuildScanEnhancements.java:
```diff
@@ -353,63 +464,56 @@ private String getGitBranchName(Supplier<String> gitCommand) {
     /**
-     * Collects custom values that should have a search link, and creates these links in `buildFinished`.
-     * The actual construction of the links must be deferred to ensure the Server URL is set.
+     * Collects all custom values that should have a search link and finally creates the links to ensure deduplication.
```
Which ones are duplicated here? And how does the de-duplication work exactly if there are multiple instances of this class?

I think the main reason that this existed was to defer the construction of the links until the server url is set. Can we be sure that all uses of this class have the server URL set already? If so do we still need this class?

jprinet replied:
Re-reading the implementation there is indeed no duplication, post-processing the navigation link is therefore useless.
I thought there were some dynamic links for a given CI server with `key` extracted from the environment but this is just not the case.
The class is still useful though to encapsulate the logic imo.

There is still one case where the server URL could not be set as `captureGitMetadata` is occurring in a `background` block which could run at configuration time depending on Gradle version.
Let me fix this.

jprinet replied:
I pushed a new commit which remove `CustomValueSearchLinker` and fixes the _server URL_ retrieval inside `captureGitMetadata`

runningcode reviewed (commented):
Getting close here! One thing, could you please add the before and after build scan links, I would also expect the outcome to be the same and that is why the links are useful.

jprinet commented:
Build scans without changes:
[Jenkins with CC stored](https://ge.solutions-team.gradle.com/s/tiwk3bu7ch4mk)
[Jenkins with CC not restored](https://ge.solutions-team.gradle.com/s/r7ujkayqyjnwq)
[Teamcity](https://ge.solutions-team.gradle.com/s/yliychbb45f4q)

Build scans with changes:
[Jenkins with CC stored](https://ge.solutions-team.gradle.com/s/b5b7h66hju5nq)
[Jenkins with CC restored](https://ge.solutions-team.gradle.com/s/qzm7fgb7o2yfe)
[Teamcity](https://ge.solutions-team.gradle.com/s/cke6ynofnh7o6)

jprinet commented:
Confirming that I was able to get CC restored in Teamcity with a custom init script.

runningcode commented on src/main/java/com/gradle/CustomBuildScanEnhancements.java:
```diff
@@ -352,75 +456,6 @@ private String getGitBranchName(Supplier<String> gitCommand) {
-                .ifPresent(x -> registerLink(linkLabel, x.getKey(), x.getValue()));
-        }
-
-        private synchronized void registerLink(String linkLabel, String name, String value) {
```
Do you know why we needed the synchronized here before? Is it safe to remove?

etiennestuder replied:
It's accessing a non-thread safe map in `customValueLinks.put` (could be invoked from multiple bg jobs in parallel).

jprinet replied:
This is not required as there is no state to keep anymore. The navigation links are added inline.

runningcode commented on src/main/java/com/gradle/CustomBuildScanEnhancements.java:
```diff
@@ -352,75 +456,6 @@ private String getGitBranchName(Supplier<String> gitCommand) {
-
-    }
-
     private void captureTestParallelization() {
```
can we move this to the bottom of the file like it used to be to lessen the diff? or is there a reason this grouping makes more logical sense?

jprinet replied:
I grouped all the static methods together at the end of the file, there was originally only 1, we could indeed move the `captureTestParallelization`  method but I find the actual setup better.

runningcode reviewed (commented):
Thanks just one comment and I think this is good to merge! Thanks for all the effort and checking everything here.

runningcode reviewed (approved):
Great work! Let's think about a release plan before merging.

jprinet commented:
Duplicate of #129, Duplicate of #130, Duplicate of #131
